monai/data/dataset.py

```diff
@@ -289,3 +286,106 @@ def __getitem__(self, index):
             # no cache for this data, execute all the transforms directly
             data = super(CacheDataset, self).__getitem__(index)
         return data
+
+
+class ZipDataset(torch.utils.data.Dataset):
+    """
+    Zip several PyTorch datasets and output data(with the same index) together in a tuple.
+    If the output of single dataset is alreay a tuple, flatten it and extend to the result.
+    For example: if datasetA returns (img, imgmeta), datsetB returns (seg, segmeta),
+    finally return (img, imgmeta, seg, segmeta).
+
+    Note:
+        Expect all the datasets have same length.
+
+    """
+
+    def __init__(self, datasets):
```

Comment:
would be useful to inherit monai.data.Dataset instead of pytorch dataset, so that it's possible to apply a transformation after the concatenation after line 323. monai's compose actually supports this type of input because of `apply_transform` https://github.com/Project-MONAI/MONAI/blob/master/monai/transforms/utils.py#L243

Reply:
I don't think we can leverage any function of MONAI Dataset, and updated the code to support `self.transform` now, could you please help review it again?